docs/source/framework/pytorch_integration/writing_layers.rst

```diff
@@ -1,137 +1,190 @@
+        # compilation and memoization
+        out, = matmul(mat1, mat2)
+        # optionally, a list of properly-sized output tensors can be passed and
+        # the kernel will use them as outputs
+        out, = matmul(mat1, mat2, outputs=[out])
+
+Specifying MappingOptions
+-----------------------------
 
-.. code-block:: python
+There are three ways to construct :code:`MappingOptions` when defining a TC:
 
-    import tensor_comprehensions as tc
-    import torch
-    MATMUL_LANG = """
-    def matmul(float(M, K) A, float(K, N) B) -> (C) {
-        C(m, n) +=! A(m, r_k) * B(r_k, n)
-    }
-    """
-    # the `name` should match the definition name in the `lang`
-    matmul = tc.define(MATMUL_LANG, name="matmul")
-    mat1, mat2 = torch.randn(3, 4).cuda(), torch.randn(4, 5).cuda()
-    out = matmul(mat1, mat2)
+* **Naive MappingOptions**:
 
-Below is a complete documentation of each API call:
+  * :code:`naive`: this is provided to create a basic mapping strategy with
+    3-D tiling by 32x32x32, mapping to 256x256 blocks and 8x32 threads. This
+    is should by no means be considered a good baseline but just a point to
+    get started using TC. Once a correct TC is written, we recommend either
+    using options loaded from a :code:`MappingOptionsCache` or resulting from
+    a tuning run.
 
-.. automodule:: tensor_comprehensions
+* **Loading from MappingOptionsCache**: a :code:`MappingOptionsCache` provides
+  a simple interface to load the best options from a previous tuning run.
 
-tc.define
----------
+* **Autotuning**: A kernel can be autotuned for fixed input tensor sizes.
+  Optionally the best performing options can be cached to a file and reused to
+  compile and run a TC operation.
 
-.. autofunction:: define
 
-.. autoclass:: TcUnit
-   :members: __call__
+Loading from cache
+------------------
 
-.. _must_pass_options:
+To load the best options from a previously saved :code:`MappingOptionsCache`
+object, one can reconstruct it explicitly from a filename and load the best
+(top-1) options given a TC string, an entry_point and a tuple of input
+tensors as such:
 
-Specifying CudaMappingOptions
------------------------------
+    .. code-block:: python
 
-TC is transformed into :code:`CUDA` kernel by using the :code:`Options` which
-is used to run the layer and hence also determines the performance of the kernel
-generated. Therefore, it is important to use good :code:`Options` for running a
-kernel. You can read more about mapping options here - :ref:`tc_mapping_options`.
+        # Setup code for mm and tensors as above
+        cache = tc.MappingOptionsCache(cache_filename)
+        best_options, = cache.load(mm, entry_point, (A, B), 1)
+        matmul = tc.define(mm, entry_point="matmul", fallback=best_options)
+        C, = matmul(A, B)
 
-There are two ways to set the :code:`Options`:
+One may also create a TC without fallback options and call compile explicitly
```

Comment:
So for this I unfortunately do not have a good answer:
1. yes there is a state, the latest compiled executor for a TC def + input sizes is the one that is memoized. I don't want to start putting protobuf + hashing on the critical path;
2. implicit compilation is made loud (instead of generating failures) and will get into tuning mode if no options are passed;
3. I don't really see a good, usable, alternative to the above for people who "just want to get a PyTorch layer working". Implicit compilation, tuning, memoization is pleasant to use but can of course come with surprises.

Let's chat more on the next PR, @apaszke your opinion would also be valuable here.